
GEDO - Topology-based inference of gene module activity in SjÃ¶grenâ€™s disease.
#ðŸ“˜ Abstract
Inferring gene module activity is key to understanding transcriptomic dysregu-
lation in case-control studies. Most existing Gene Set Analysis methods focus on
the relative positioning of data points without considering the complex geome-
try of the data.
We introduce GEDO, a graph- and topology-based method that infers gene
module activity through a transition score quantifying the shift from healthy
controls to diseased individuals. When applied to bulk RNA-seq data from
SjÂ¨ogrenâ€™s disease patients and healthy controls (PRECISESADS cohort), GEDO
was benchmarked against PCA1 and the Mean of z-scores in classification and
unsupervised clustering tasks.
GEDO outperformed both approaches in predicting disease status and yielded
more informative module matrices, enhancing subgroup separability. The bio-
logical signal captured was aligned with the knowledge and clinical features of
SjÂ¨ogrenâ€™s disease.
GEDO offers a robust, interpretable framework for quantifying gene modules
activity, with applications in single- and multi-omics integration. Its topology-
based design enables finer resolution of disease-related transcriptomic alterations.

#ðŸ“‚ Structure of the repository
1. TUTORIAL.html : Quarto edited file which run analysis, show code execution, and show figures.
2. functions_article.R : functions for computing module matrices with all methods exepts GEDO, and ton run comparison analyses.
3. GEDO_toy_example.R : script which run GEDO on random values generated directly in the script, and plot a heatmap. 
4. GEDO.R : not provided in this repository. Contain functions necessary to run GEDO algorithm. This script can be shared upon request to the authors, and under certain conditions. 

# ðŸ“„ Tutorial 
ðŸ‘‰ [TUTORIAL.html](TUTORIAL.html)


